csaf-validation/src/main/kotlin/io/github/csaf/sbom/validation/tests/Tests.kt

```diff
@@ -0,0 +1,289 @@
+        Test613CircularDefinitionOfProductID,
+        Test614MissingDefinitionOfProductGroupID,
+        Test615MultipleDefinitionOfProductGroupID,
+        Test616ContradictingProductStatus,
+        Test617MultipleScoresWithSameVersionPerProduct,
+    )
+
+/**
+ * Optional tests as defined in
+ * [Section 6.2](https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#62-optional-tests).
+ */
+var optionalTests =
+    listOf(
+        Test621UnusedDefinitionOfProductID,
+    )
+
+/**
+ * Informative tests as defined in
+ * [Section 6.3](https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#63-informative-test).
+ */
+var informativeTests = listOf<Test>()
+
+/** Executes all tests in this list of [Test] objects. */
+fun List<Test>.test(doc: Csaf): ValidationResult {
+    return this.map { it.test(doc) }.merge()
+}
+
+/**
+ * Implementation of
+ * [Test 6.1.1](https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#611-missing-definition-of-product-id).
+ */
+object Test611MissingDefinitionOfProductID : Test {
+    override fun test(doc: Csaf): ValidationResult {
+        val definitions = doc.gatherProducts().map { it.product_id }
+        val references = doc.gatherProductReferences()
+
+        val notDefined = references.subtract(definitions.toSet())
+
+        return if (notDefined.isEmpty()) {
+            ValidationSuccessful
+        } else {
+            ValidationFailed(
+                listOf("The following IDs are not defined: ${notDefined.joinToString(",")}")
+            )
+        }
+    }
+}
+
+/**
+ * Implementation of
+ * [Test 6.1.2](https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#612-multiple-definition-of-product-id).
+ */
+object Test612MultipleDefinitionOfProductID : Test {
+    override fun test(doc: Csaf): ValidationResult {
+        val definitions = doc.gatherProducts().map { it.product_id }
+
+        val duplicates = definitions.duplicates()
+
+        return if (duplicates.isEmpty()) {
+            ValidationSuccessful
+        } else {
+            ValidationFailed(
+                listOf("The following IDs are duplicate: ${duplicates.keys.joinToString(",")}")
+            )
+        }
+    }
+}
+
+private fun <T> List<T>.duplicates(): Map<T, Int> {
+    return groupingBy { it }.eachCount().filter { it.value > 1 }
+}
+
+/**
+ * Implementation of
+ * [Test 6.1.3](https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#613-circular-definition-of-product-id).
+ */
+object Test613CircularDefinitionOfProductID : Test {
+    override fun test(doc: Csaf): ValidationResult {
+        val circles = mutableSetOf<String>()
+
+        for (relationship in doc.product_tree?.relationships ?: listOf()) {
+            val definedId = relationship.full_product_name.product_id
+            val notAllowed =
+                listOf(relationship.product_reference, relationship.relates_to_product_reference)
+            if (definedId in notAllowed) {
+                circles += definedId
```

Comment:
Do we really want to keep iterating even if it's clear that the test fails? Of course, it gives us more information but there might be some performance penalty.

Reply:
I would say: yes. It it will be helpful if there is more than one failure, we do it also similarly in the other tests.

Reply:
Fine for me